File: v1/src/main/java/com/api/v1/services/creditcard/AllCreditCardsRetrievalServiceImpl.java
package com.api.v1.services.creditcard;

import com.api.v1.annotations.creditcard.CardNumber;
import com.api.v1.domain.creditcard.CreditCardRepository;
import com.api.v1.dtos.creditcard.CreditCardResponseDto;
import com.api.v1.exceptions.creditcard.EmptyCreditCardEntityException;
import com.api.v1.mappers.creditcard.CreditCardResponseMapper;
import com.api.v1.utils.creditcard.CreditCardFinderUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
class AllCreditCardsRetrievalServiceImpl implements AllCreditCardsRetrievalService {

    @Autowired
    private CreditCardRepository repository;

    @Autowired
    private CreditCardFinderUtil creditCardFinderUtil;

    @Override
    public Flux<CreditCardResponseDto> findAll() {
        return repository
                .findAll()
                .hasElements()
                .flatMapMany(hasElements -> {
                    if (!hasElements) return Mono.error(new EmptyCreditCardEntityException());
                    return repository
                            .findAll()
                            .flatMap(e -> Flux.just(new CreditCardResponseMapper(e).mapToDto()));
                });
    }

    @Override
    public Mono<CreditCardResponseDto> findByCardNumber(@CardNumber String cardNumber) {
        return creditCardFinderUtil
                .find(cardNumber)
                .flatMap(e -> Mono.just(new CreditCardResponseMapper(e).mapToDto()));
    }

}

File: v2/src/main/kotlin/com/api/v2/creditcard/services/AllCreditCardsRetrievalServiceImpl.kt
package com.api.v2.creditcard.services;

import com.api.v2.annotations.EmptyFlowException
import com.api.v2.creditcard.domain.CreditCardRepository
import com.api.v2.creditcard.dtos.CreditCardResponseDto
import com.api.v2.creditcard.mappers.CreditCardResponseMapper
import com.api.v2.creditcard.utils.CreditCardFinderUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service;

@Service
private class AllCreditCardsRetrievalServiceImpl: AllCreditCardsRetrievalService {

    @Autowired
    lateinit var creditCardFinderUtil: CreditCardFinderUtil

    @Autowired
    lateinit var creditCardRepository: CreditCardRepository

    override suspend  fun findByCardNumber(cardNumber: String): CreditCardResponseDto {
        return withContext(Dispatchers.IO) {
            val creditCard = creditCardFinderUtil.find(cardNumber)
            CreditCardResponseMapper.map(creditCard)
        }
    }

    override suspend  fun findAll(): Flow<CreditCardResponseDto> {
        return withContext(Dispatchers.IO) {
            if (creditCardRepository.findAll().count() == 0) {
                throw EmptyFlowException()
            }
            creditCardRepository.findAll().map { e -> CreditCardResponseMapper.map(e) }
        }
    }

}
